File: ta4j-core/src/test/java/org/ta4j/core/indicators/numeric/NumericIndicatorTest.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.indicators.numeric;

import static org.junit.Assert.assertEquals;
import static org.ta4j.core.TestUtils.assertNumEquals;

import org.junit.Test;
import org.ta4j.core.BarSeries;
import org.ta4j.core.Indicator;
import org.ta4j.core.indicators.AbstractIndicatorTest;
import org.ta4j.core.indicators.adx.ADXIndicator;
import org.ta4j.core.indicators.averages.EMAIndicator;
import org.ta4j.core.indicators.averages.SMAIndicator;
import org.ta4j.core.indicators.helpers.ClosePriceIndicator;
import org.ta4j.core.indicators.helpers.HighestValueIndicator;
import org.ta4j.core.indicators.helpers.LowestValueIndicator;
import org.ta4j.core.indicators.helpers.VolumeIndicator;
import org.ta4j.core.indicators.statistics.StandardDeviationIndicator;
import org.ta4j.core.mocks.MockBarSeriesBuilder;
import org.ta4j.core.num.NaN;
import org.ta4j.core.num.Num;
import org.ta4j.core.num.NumFactory;
import org.ta4j.core.rules.CrossedDownIndicatorRule;
import org.ta4j.core.rules.CrossedUpIndicatorRule;
import org.ta4j.core.rules.OverIndicatorRule;
import org.ta4j.core.rules.UnderIndicatorRule;

public class NumericIndicatorTest extends AbstractIndicatorTest<NumericIndicator, Num> {

    private final BarSeries series = new MockBarSeriesBuilder().withNumFactory(numFactory)
            .withData(1, 2, 3, 4, 5, 6, 7, 8, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2)
            .build();
    private final ClosePriceIndicator cp1 = new ClosePriceIndicator(series);
    private final EMAIndicator ema = new EMAIndicator(cp1, 3);

    public NumericIndicatorTest(NumFactory numFactory) {
        super(numFactory);
    }

    @Test
    public void plus() {
        final NumericIndicator numericIndicator = NumericIndicator.of(cp1);

        final NumericIndicator staticOp = numericIndicator.plus(5);
        assertNumEquals(1 + 5, staticOp.getValue(0));
        assertNumEquals(9 + 5, staticOp.getValue(8));

        final NumericIndicator dynamicOp = numericIndicator.plus(ema);
        assertNumEquals(cp1.getValue(0).plus(ema.getValue(0)), dynamicOp.getValue(0));
        assertNumEquals(cp1.getValue(8).plus(ema.getValue(8)), dynamicOp.getValue(8));
    }

    @Test
    public void minus() {
        final NumericIndicator numericIndicator = NumericIndicator.of(cp1);

        final NumericIndicator staticOp = numericIndicator.minus(5);
        assertNumEquals(1 - 5, staticOp.getValue(0));
        assertNumEquals(9 - 5, staticOp.getValue(8));

        final NumericIndicator dynamicOp = numericIndicator.minus(ema);
        assertNumEquals(cp1.getValue(0).minus(ema.getValue(0)), dynamicOp.getValue(0));
        assertNumEquals(cp1.getValue(8).minus(ema.getValue(8)), dynamicOp.getValue(8));
    }

    @Test
    public void multipliedBy() {
        final NumericIndicator numericIndicator = NumericIndicator.of(cp1);

        final NumericIndicator staticOp = numericIndicator.multipliedBy(5);
        assertNumEquals(1 * 5, staticOp.getValue(0));
        assertNumEquals(9 * 5, staticOp.getValue(8));

        final NumericIndicator dynamicOp = numericIndicator.multipliedBy(ema);
        assertNumEquals(cp1.getValue(0).multipliedBy(ema.getValue(0)), dynamicOp.getValue(0));
        assertNumEquals(cp1.getValue(8).multipliedBy(ema.getValue(8)), dynamicOp.getValue(8));
    }

    @Test
    public void dividedBy() {
        final NumericIndicator numericIndicator = NumericIndicator.of(cp1);

        final NumericIndicator staticOp = numericIndicator.dividedBy(5);
        assertNumEquals(1 / 5.0, staticOp.getValue(0));
        assertNumEquals(9 / 5.0, staticOp.getValue(8));

        final NumericIndicator zeroOp = numericIndicator.dividedBy(0);
        assertNumEquals(NaN.NaN, zeroOp.getValue(0));
        assertNumEquals(NaN.NaN, zeroOp.getValue(8));

        final NumericIndicator dynamicOp = numericIndicator.dividedBy(ema);
        assertNumEquals(cp1.getValue(0).dividedBy(ema.getValue(0)), dynamicOp.getValue(0));
        assertNumEquals(cp1.getValue(8).dividedBy(ema.getValue(8)), dynamicOp.getValue(8));
    }

    @Test
    public void max() {
        final NumericIndicator numericIndicator = NumericIndicator.of(cp1);

        final NumericIndicator staticOp = numericIndicator.max(5);
        assertNumEquals(5, staticOp.getValue(0));
        assertNumEquals(9, staticOp.getValue(8));

        final NumericIndicator dynamicOp = numericIndicator.max(ema);
        assertNumEquals(cp1.getValue(0).max(ema.getValue(0)), dynamicOp.getValue(0));
        assertNumEquals(cp1.getValue(8).max(ema.getValue(8)), dynamicOp.getValue(8));
    }

    @Test
    public void min() {
        final NumericIndicator numericIndicator = NumericIndicator.of(cp1);

        final NumericIndicator staticOp = numericIndicator.min(5);
        assertNumEquals(1, staticOp.getValue(0));
        assertNumEquals(5, staticOp.getValue(8));

        final NumericIndicator dynamicOp = numericIndicator.min(ema);
        assertNumEquals(cp1.getValue(0).min(ema.getValue(0)), dynamicOp.getValue(0));
        assertNumEquals(cp1.getValue(8).min(ema.getValue(8)), dynamicOp.getValue(8));
    }

    @Test
    public void sqrt() {
        final NumericIndicator numericIndicator = NumericIndicator.of(cp1);
        final NumericIndicator dynamicOp = numericIndicator.sqrt();
        assertNumEquals(1, dynamicOp.getValue(0));
        assertNumEquals(Math.sqrt(2.0), dynamicOp.getValue(1));
        assertNumEquals(3, dynamicOp.getValue(8));
    }

    @Test
    public void abs() {
        final NumericIndicator numericIndicator = NumericIndicator.of(cp1);
        final NumericIndicator dynamicOp = numericIndicator.abs();
        assertNumEquals(1, dynamicOp.getValue(0));
        assertNumEquals(2, dynamicOp.getValue(series.getBarCount() - 1));
    }

    @Test
    public void squared() {
        final NumericIndicator numericIndicator = NumericIndicator.of(cp1);
        final NumericIndicator dynamicOp = numericIndicator.squared();
        assertNumEquals(1, dynamicOp.getValue(0));
        assertNumEquals(81, dynamicOp.getValue(8));
    }

    @Test
    public void indicators() {
        final NumericIndicator numericIndicator = NumericIndicator.of(cp1);

        assertEquals(SMAIndicator.class, numericIndicator.sma(5).delegate().getClass());
        assertEquals(EMAIndicator.class, numericIndicator.ema(5).delegate().getClass());
        assertEquals(StandardDeviationIndicator.class, numericIndicator.stddev(5).delegate().getClass());
        assertEquals(HighestValueIndicator.class, numericIndicator.highest(5).delegate().getClass());
        assertEquals(LowestValueIndicator.class, numericIndicator.lowest(5).delegate().getClass());

        assertEquals(ClosePriceIndicator.class, NumericIndicator.closePrice(series).delegate().getClass());
        assertEquals(VolumeIndicator.class, NumericIndicator.volume(series).delegate().getClass());

        ADXIndicator adx = new ADXIndicator(series, 5);
        assertEquals(ADXIndicator.class, NumericIndicator.of(adx).delegate().getClass());
    }

    @Test
    public void rules() {
        final NumericIndicator numericIndicator = NumericIndicator.of(cp1);

        assertEquals(CrossedUpIndicatorRule.class, numericIndicator.crossedOver(5).getClass());
        assertEquals(CrossedUpIndicatorRule.class, numericIndicator.crossedOver(ema).getClass());
        assertEquals(CrossedDownIndicatorRule.class, numericIndicator.crossedUnder(5).getClass());
        assertEquals(CrossedDownIndicatorRule.class, numericIndicator.crossedUnder(ema).getClass());
        assertEquals(OverIndicatorRule.class, numericIndicator.isGreaterThan(5).getClass());
        assertEquals(OverIndicatorRule.class, numericIndicator.isGreaterThan(ema).getClass());
        assertEquals(UnderIndicatorRule.class, numericIndicator.isLessThan(5).getClass());
        assertEquals(UnderIndicatorRule.class, numericIndicator.isLessThan(ema).getClass());
    }

    @Test
    public void previous() {
        final NumericIndicator numericIndicator = NumericIndicator.of(cp1);
        final Indicator<Num> previous = numericIndicator.previous();
        assertNumEquals(cp1.getValue(0), previous.getValue(1));

        final Indicator<Num> previous3 = numericIndicator.previous(3);
        assertNumEquals(cp1.getValue(3), previous3.getValue(6));
    }

    @Test
    public void getValue() {
        final NumericIndicator numericIndicator = NumericIndicator.of(cp1);

        for (int i = 0; i < series.getBarCount(); i++) {
            assertNumEquals(cp1.getValue(i), numericIndicator.getValue(i));
        }
    }

    @Test
    public void barSeries() {
        final NumericIndicator numericIndicator = NumericIndicator.of(cp1);
        assertEquals(cp1.getBarSeries(), numericIndicator.getBarSeries());
    }
}

File: ta4j-core/src/test/kotlin/org/ta4j/core/indicators/numeric/NumericIndicatorTest.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.indicators.numeric

import org.junit.Assert
import org.junit.Test
import org.ta4j.core.BarSeries
import org.ta4j.core.Indicator
import org.ta4j.core.TestUtils
import org.ta4j.core.indicators.AbstractIndicatorTest
import org.ta4j.core.indicators.EMAIndicator
import org.ta4j.core.indicators.SMAIndicator
import org.ta4j.core.indicators.adx.ADXIndicator
import org.ta4j.core.indicators.helpers.ClosePriceIndicator
import org.ta4j.core.indicators.helpers.HighestValueIndicator
import org.ta4j.core.indicators.helpers.LowestValueIndicator
import org.ta4j.core.indicators.helpers.VolumeIndicator
import org.ta4j.core.indicators.numeric.NumericIndicator.Companion.closePrice
import org.ta4j.core.indicators.numeric.NumericIndicator.Companion.of
import org.ta4j.core.indicators.numeric.NumericIndicator.Companion.volume
import org.ta4j.core.indicators.statistics.StandardDeviationIndicator
import org.ta4j.core.mocks.MockBarSeries
import org.ta4j.core.num.NaN
import org.ta4j.core.num.Num
import org.ta4j.core.rules.CrossedDownIndicatorRule
import org.ta4j.core.rules.CrossedUpIndicatorRule
import org.ta4j.core.rules.OverIndicatorRule
import org.ta4j.core.rules.UnderIndicatorRule
import java.util.function.Function

class NumericIndicatorTest(numFunction: Function<Number?, Num>) :
    AbstractIndicatorTest<NumericIndicator?, Num>(numFunction) {
    private val series: BarSeries = MockBarSeries(
        numFunction, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 8.0, 7.0, 6.0, 5.0, 4.0, 3.0, 2.0, 1.0,
        0.0, -1.0, -2.0
    )
    private val cp1 = ClosePriceIndicator(series)
    private val ema = EMAIndicator(cp1, 3)
    @Test
    fun plus() {
        val numericIndicator = of(cp1)
        val staticOp = numericIndicator.plus(5)
        TestUtils.assertNumEquals(1 + 5, staticOp[0])
        TestUtils.assertNumEquals(9 + 5, staticOp.getValue(8))
        val dynamicOp = numericIndicator.plus(ema)
        TestUtils.assertNumEquals(cp1[0].plus(ema[0]), dynamicOp[0])
        TestUtils.assertNumEquals(cp1.getValue(8).plus(ema.getValue(8)), dynamicOp.getValue(8))
    }

    @Test
    fun minus() {
        val numericIndicator = of(cp1)
        val staticOp = numericIndicator.minus(5)
        TestUtils.assertNumEquals(1 - 5, staticOp[0])
        TestUtils.assertNumEquals(9 - 5, staticOp.getValue(8))
        val dynamicOp = numericIndicator.minus(ema)
        TestUtils.assertNumEquals(cp1[0].minus(ema[0]), dynamicOp[0])
        TestUtils.assertNumEquals(cp1.getValue(8).minus(ema.getValue(8)), dynamicOp.getValue(8))
    }

    @Test
    fun multipliedBy() {
        val numericIndicator = of(cp1)
        val staticOp = numericIndicator.times(5)
        TestUtils.assertNumEquals(1 * 5, staticOp[0])
        TestUtils.assertNumEquals(9 * 5, staticOp.getValue(8))
        val dynamicOp = numericIndicator.times(ema)
        TestUtils.assertNumEquals(cp1[0].times(ema[0]), dynamicOp[0])
        TestUtils.assertNumEquals(cp1.getValue(8).times(ema.getValue(8)), dynamicOp.getValue(8))
    }

    @Test
    fun dividedBy() {
        val numericIndicator = of(cp1)
        val staticOp = numericIndicator.div(5)
        TestUtils.assertNumEquals(1 / 5.0, staticOp[0])
        TestUtils.assertNumEquals(9 / 5.0, staticOp.getValue(8))
        val zeroOp = numericIndicator.div(0)
        TestUtils.assertNumEquals(NaN.NaN, zeroOp[0])
        TestUtils.assertNumEquals(NaN.NaN, zeroOp.getValue(8))
        val dynamicOp = numericIndicator.div(ema)
        TestUtils.assertNumEquals(cp1[0].div(ema[0]), dynamicOp[0])
        TestUtils.assertNumEquals(cp1.getValue(8).div(ema.getValue(8)), dynamicOp.getValue(8))
    }

    @Test
    fun max() {
        val numericIndicator = of(cp1)
        val staticOp = numericIndicator.max(5)
        TestUtils.assertNumEquals(5, staticOp[0])
        TestUtils.assertNumEquals(9, staticOp.getValue(8))
        val dynamicOp = numericIndicator.max(ema)
        TestUtils.assertNumEquals(cp1[0].max(ema[0]), dynamicOp[0])
        TestUtils.assertNumEquals(cp1.getValue(8).max(ema.getValue(8)), dynamicOp.getValue(8))
    }

    @Test
    fun min() {
        val numericIndicator = of(cp1)
        val staticOp = numericIndicator.min(5)
        TestUtils.assertNumEquals(1, staticOp[0])
        TestUtils.assertNumEquals(5, staticOp.getValue(8))
        val dynamicOp = numericIndicator.min(ema)
        TestUtils.assertNumEquals(cp1[0].min(ema[0]), dynamicOp[0])
        TestUtils.assertNumEquals(cp1.getValue(8).min(ema.getValue(8)), dynamicOp.getValue(8))
    }

    @Test
    fun sqrt() {
        val numericIndicator = of(cp1)
        val dynamicOp = numericIndicator.sqrt()
        TestUtils.assertNumEquals(1, dynamicOp[0])
        TestUtils.assertNumEquals(Math.sqrt(2.0), dynamicOp.getValue(1))
        TestUtils.assertNumEquals(3, dynamicOp.getValue(8))
    }

    @Test
    fun abs() {
        val numericIndicator = of(cp1)
        val dynamicOp = numericIndicator.abs()
        TestUtils.assertNumEquals(1, dynamicOp[0])
        TestUtils.assertNumEquals(2, dynamicOp.getValue(series.barCount - 1))
    }

    @Test
    fun squared() {
        val numericIndicator = of(cp1)
        val dynamicOp = numericIndicator.squared()
        TestUtils.assertNumEquals(1, dynamicOp[0])
        TestUtils.assertNumEquals(81, dynamicOp.getValue(8))
    }

    @Test
    fun indicators() {
        val numericIndicator = of(cp1)
        Assert.assertEquals(SMAIndicator::class.java, numericIndicator.sma(5).delegate().javaClass)
        Assert.assertEquals(EMAIndicator::class.java, numericIndicator.ema(5).delegate().javaClass)
        Assert.assertEquals(StandardDeviationIndicator::class.java, numericIndicator.stddev(5).delegate().javaClass)
        Assert.assertEquals(HighestValueIndicator::class.java, numericIndicator.highest(5).delegate().javaClass)
        Assert.assertEquals(LowestValueIndicator::class.java, numericIndicator.lowest(5).delegate().javaClass)
        Assert.assertEquals(ClosePriceIndicator::class.java, closePrice(series).delegate().javaClass)
        Assert.assertEquals(VolumeIndicator::class.java, volume(series).delegate().javaClass)
        val adx = ADXIndicator(series, 5)
        Assert.assertEquals(ADXIndicator::class.java, of(adx).delegate().javaClass)
    }

    @Test
    fun rules() {
        val numericIndicator = of(cp1)
        Assert.assertEquals(CrossedUpIndicatorRule::class.java, numericIndicator.crossedOver(5).javaClass)
        Assert.assertEquals(CrossedUpIndicatorRule::class.java, numericIndicator.crossedOver(ema).javaClass)
        Assert.assertEquals(CrossedDownIndicatorRule::class.java, numericIndicator.crossedUnder(5).javaClass)
        Assert.assertEquals(CrossedDownIndicatorRule::class.java, numericIndicator.crossedUnder(ema).javaClass)
        Assert.assertEquals(OverIndicatorRule::class.java, numericIndicator.isGreaterThan(5).javaClass)
        Assert.assertEquals(OverIndicatorRule::class.java, numericIndicator.isGreaterThan(ema).javaClass)
        Assert.assertEquals(UnderIndicatorRule::class.java, numericIndicator.isLessThan(5).javaClass)
        Assert.assertEquals(UnderIndicatorRule::class.java, numericIndicator.isLessThan(ema).javaClass)
    }

    @Test
    fun previous() {
        val numericIndicator = of(cp1)
        val previous = numericIndicator.previous()
        TestUtils.assertNumEquals(cp1[0], previous.getValue(1))
        val previous3: Indicator<Num> = numericIndicator.previous(3)
        TestUtils.assertNumEquals(cp1.getValue(3), previous3.getValue(6))
    }

    @Test
    fun getValue() {
        val numericIndicator = of(cp1)
        for (i in 0 until series.barCount) {
            TestUtils.assertNumEquals(cp1[i], numericIndicator[i])
        }
    }

    @Test
    fun barSeries() {
        val numericIndicator = of(cp1)
        Assert.assertEquals(cp1.barSeries, numericIndicator.barSeries)
    }

    @Test
    fun numOf() {
        val numericIndicator = of(cp1)
        TestUtils.assertNumEquals(cp1.numOf(0), numericIndicator.numOf(0))
    }
}